src/librustdoc/doctree.rs

```diff
@@ -29,6 +59,35 @@ impl Module<'hir> {
             foreigns: Vec::new(),
             macros: Vec::new(),
             is_crate: false,
+            from_glob: false,
+        }
+    }
+
+    pub(crate) fn push_item(&mut self, new_item: Item<'hir>) {
+        if let Some(existing_item) =
+            self.items.iter_mut().find(|item| item.name() == new_item.name())
```

Comment:
If `new_item.name().is_empty()` then it should always be added. This applies to items like impls and glob imports (that aren't inlined). They can't shadow each other and it doesn't make sense to compare the names.

Also this will need to check if the namespaces match as well. For example:

```rust
mod foo {
    pub struct Foo;
}

pub use foo::*;

pub fn Foo() {}
```

Both the struct `Foo` and the function `Foo` should appear in docs because they are in different namespaces.

Reply:
> Both the struct Foo and the function Foo should appear in docs because they are in different namespaces.

If you use `modname::Foo` in rust code, it will always resolves to the `Foo` function and never resolves to the struct, so IMO the function `Foo` do shadows the struct `Foo`, so it is reasonable for the doc to do the same thing.

Do you have any other considerations?